### DISCUSSION TITLE
Who adds `.system()` to `hello_world.system()`?

### DISCUSSION THREAD

**Discussion by abitrolly** [bevyengine/bevy#1137] [Q&A] [answered]
I am reading https://bevyengine.org/learn/book/getting-started/ecs/#your-first-system which contains the following code.
```rust
use bevy::prelude::*;

fn hello_world() {
    println!("hello world!");
}

fn main() {
    App::build()
        .add_system(hello_world.system())
        .run();
}
```
I don't get where the `system()` method got attached to `hello_world`. As I am not a Rust programmer, the explanation that this is "trait extension method" doesn't tell me anything. It would be nice to expand where exactly that attachment happens.

**Comment by DJMcNab** [comment]
It's an implementation of the `bevy_ecs::system::IntoSystem` trait:
https://github.com/bevyengine/bevy/blob/61ce3f7bcfbff808f2eb51e914f03d3807a8f92a/crates/bevy_ecs/src/system/into_system.rs#L193-L195

The implementation specifically for `hello_world` is generated in this macro call.

https://github.com/bevyengine/bevy/blob/61ce3f7bcfbff808f2eb51e914f03d3807a8f92a/crates/bevy_ecs/src/system/into_system.rs#L314

And the actual implementation is generated here, because that function does not have an `Input` parameter.

https://github.com/bevyengine/bevy/blob/61ce3f7bcfbff808f2eb51e914f03d3807a8f92a/crates/bevy_ecs/src/system/into_system.rs#L217

**Comment by abitrolly** [comment]
@DJMcNab does that mean that all functions in the file get the `.system()` method?

**Reply by DJMcNab** [reply]
Any function which implements the `IntoSystem` trait would. That would be any function where all of up to 16 parameters implement `SystemParam`, or the first parameter is `In<SomeType>`, and then has up to 16 parameters which implement `SystemParam`. 
`SystemParam` is defined as:
https://github.com/bevyengine/bevy/blob/61ce3f7bcfbff808f2eb51e914f03d3807a8f92a/crates/bevy_ecs/src/system/system_param.rs#L8-L10
The types which implement `SystemParam` are listed here <https://docs.rs/bevy_ecs/0.4.0/bevy_ecs/trait.SystemParam.html#implementors>, and it is also implemented on tuples where every member is SystemParam, again up to 16 elements.

**Reply by abitrolly** [reply]
Coming from Python I still don't get why it works. The definition for `hello_world()` doesn't include any params, so no `SystemParam` reference anywhere.
```rust
fn hello_world() {
```
When Rust decides that `hello_world()` implements this `IntoSystem`? Is it when Rust reads the text file and encounters `hello_world()`? Or maybe it decided later - when all files are read, then some macro processor goes through all global functions and monkey patches them?

**Reply by smokku** [reply]
Simplifying things a bit.

In Bevy there is a code like this:
```rust
impl IntoSystem for FnMut()
{
    fn system(mut self) -> FuncSystem {
        ...
    }
}
```
Which implements `IntoSystem` for a function without parameters (`FnMut()`).

When compiler sees `hello_world.system()` it does something like this:
1. Ok - the dot means I need to call a method.
2. What is the type of the thing left to dot (`hello_world`)? It is `FnMut()`.
3. Do I have any trait in scope that defines what is the right of the dot (the `system()` method) on `FnMut()`? Yes - the `IntoSystem` trait.
4. Do I have an implementation of that trait available? Yes (see above).
5. Ok. Let's generate a call to that function passing it the thing on the left of dot.

This way `hello_world` function becomes `mut self` inside the `system()` function call, gets wrapped in a new `FuncSystem` and passed to `.add_system()`.

That's basically it on the simplest level.
All the macro shenanigans is just so we do not have to write `impl IntoSystem for FnMut( ....` for every possible set of parameters we want to support. There is a macro that generates all these.

**Reply by DJMcNab** [reply]
The way this gets resolved, is that the compiler takes the reference to `hello_world`, which is a function point, having type `fn()->()`. It then looks for any traits implemented for `fn()->()`, which includes `IntoSystem`, because `fn()->()` meets all of the bounds for the `impl` of `IntoSystem<(), FuncSystem<()>>`.

This is the implementation:
https://github.com/bevyengine/bevy/blob/61ce3f7bcfbff808f2eb51e914f03d3807a8f92a/crates/bevy_ecs/src/system/into_system.rs#L217
which is in the expansion of:
https://github.com/bevyengine/bevy/blob/61ce3f7bcfbff808f2eb51e914f03d3807a8f92a/crates/bevy_ecs/src/system/into_system.rs#L314
The expanded impl is:
```rust
impl<Func, Out> IntoSystem<(), FuncSystem<Out>> for Func
    where Func: FnMut()->() + FnMut()->Out + Send + Sync + 'static, Out: 'static {
    fn system(mut self)->FuncSystem<Out> {...}
}
```

It then calls the `system` function from that implementation, with `self` as the function pointer value `hello_world`

**Comment by TarekAS** [comment]
- Every function in Rust is a variable, it has a name, type, and a value.
- The name of the variable is the function name, the type is the function signature, and the value is the code block. We don't care about the code block.
- We can implement traits (i.e. interfaces) for any type, including function signatures.
- ```IntoSystem``` is a trait that is generic over all function signatures. Which means that ```fn hello_world()``` is eligible to implement it.
- To implement the trait, you must implement a ```.system()``` method on the function. As @smokku shows above.
- Because we don't care about the code block of the function, the ```.system()``` logic we need to implement is identical for all functions that have the same function signature (i.e. type).
- Therefore, we can automate this process. This is done for your by Bevy through the ```impl_into_system!();``` macro, which generates the ```.system()``` implementation for your function signature at compile time.
- ```.system()``` is a method that returns your function wrapped up in a ```FuncSystem``` type, which is required as input to ```.add_system```.

Obviously I'm simplifying things a bit (and I'm a Rust newbie myself), but hopefully it makes things more understandable.

**Reply by abitrolly** [reply]
Thanks. That clarifies how trait is matched against `hello_world` function and what the `.system()` call returns. What is not completely clear is the distinction between trait definition and implementation.

If this is definition.
```rust
impl IntoSystem for FnMut()
{
    fn system(mut self) -> FuncSystem {
        ...
    }
}
```
Then I don't understand why additional implementation is needed. Because this seems to contain all the details needed to match the function signature (`hello_world()` matches `FnMut()`) and add a method to it.

**Reply by DJMcNab** [reply]
That's required because of the orphan rules - you can't implement non-trait methods on a type unless you are in the same crate as it is defined. And `bevy_ecs` is not `{inside the compiler}`, which is where the type of hello world, which is `fn()` is defined. Hence a trait is required to add the `.system()` method.

**Reply by smokku** [reply]
> Then I don't understand why additional implementation is needed. Because this seems to contain all the details needed to match the function signature (`hello_world()` matches `FnMut()`) and add a method to it.

This covers `fn hello_world()` system.
But what about `fn move_players(players: Query<&Player>)` system? Or `fn tick_timer(timer: ResMut<MyTimer>, time: Res<Time>)` system?

These require `impl IntoSystem<T> for FnMut(Query<T>) { ...` and `impl IntoSystem<T, R> for FnMut(ResMut<T>, Res<R>) { ...`.
Do you want to write all these implementations manually? If not, use macros. :smiley: 

Of course I'm simplifying for illustration. Real `IntoSystem` implementation is generic over traits. But argument count and type for `Fn` must match - we cannot generalize it.

**Reply by abitrolly** [reply]
Is there a distinction between trait definition and implementation? Quoting @smokku.

> 3. Do I have any trait in scope that defines what is the right of the dot (the system() method) on FnMut()? Yes - the IntoSystem trait.
> 4. Do I have an implementation of that trait available? Yes (see above).

If I understand it correctly, the 3rd party crates can not monkey patch code my crate. That's fair. But 3rd party cargo can define a trait that I can reuse. The implementation is the way to import the trait from another crate. Is that right?

**Reply by abitrolly** [reply]
The macro expansion part is quite clear. )

**Reply by smokku** [reply]
> If I understand it correctly, the 3rd party crates can not monkey patch code my crate.

Rust is not a dynamic language running on a virtual machine. Code is compiled to native binary and runs from read-only memory.
Monkey-patching existing code is not possible.

3rd party crates can provide functions and your code can call these functions.
3rd party crates can also provide templates for your application on how to generate code depending on generic type parameters. These generated functions become part of your app.

> Is there a distinction between trait definition and implementation?

Trait definition is a way of saying "The type implementing this trait will have such methods". It is also a way of saying "I will take a type implementing these methods as an argument". You can import Trait definitions from other crates and define own in own crate.

Trait implementation is an actual code providing the instructions for a method on a specific type. You can import Trait implementations from other crates and provide own in own crate.

**Reply by abitrolly** [reply]
There are more details than fits in my head, but that's already enough to move on. Thanks for replies! :D

**Reply by abitrolly** [reply]
I would mark https://github.com/bevyengine/bevy/discussions/1137#discussioncomment-237471 as an answer, but GitHub doesn't allow me to do so.

**Reply by DJMcNab** [reply]
@smokku if you make a top level answer that's a clone of that link, then that can be marked as answer

**Reply by smokku** [reply]
Like so?
https://github.com/bevyengine/bevy/discussions/1137#discussioncomment-238609

**Reply by DJMcNab** [reply]
Yeah, exactly

**Comment by smokku** [comment, marked as answer]
Simplifying things a bit.

In Bevy there is a code like this:
```rust
impl IntoSystem for FnMut()
{
    fn system(mut self) -> FuncSystem {
        ...
    }
}
```
Which implements `IntoSystem` for a function without parameters (`FnMut()`).

When compiler sees `hello_world.system()` it does something like this:
1. Ok - the dot means I need to call a method.
2. What is the type of the thing left to dot (`hello_world`)? It is `FnMut()`.
3. Do I have any trait in scope that defines what is the right of the dot (the `system()` method) on `FnMut()`? Yes - the `IntoSystem` trait.
4. Do I have an implementation of that trait available? Yes (see above).
5. Ok. Let's generate a call to that function passing it the thing on the left of dot.

This way `hello_world` function becomes `mut self` inside the `system()` function call, gets wrapped in a new `FuncSystem` and passed to `.add_system()`.

That's basically it on the simplest level.
All the macro shenanigans is just so we do not have to write `impl IntoSystem for FnMut( ....` for every possible set of parameters we want to support. There is a macro that generates all these.
